tenant/storage.go

```diff
@@ -0,0 +1,115 @@
+package tenant
+
+import (
+	"context"
+	"errors"
+	"sync"
+
+	"github.com/influxdata/influxdb/kv"
+)
+
+var errRollback = errors.New("rollback")
+
+type Store struct {
+	kvStore kv.Store
+}
+
+type TxType int
+
+const (
+	Write TxType = iota
+	Read
+)
+
+type Tx struct {
+	kv.Tx
+	close    chan error // Commit() closes the channel, and Rollback sends a rollback error on the channel
+	response chan error // used to get the response from the transaction handling go routine
+}
+
+func (tx *Tx) Commit() error {
+	close(tx.close)
+	return <-tx.response
+}
+
+func (tx *Tx) Rollback() error {
+	tx.close <- errRollback
+	close(tx.close)
+	err := <-tx.response
+	if err != errRollback {
+		return err
+	}
+	return nil
+}
+
+func NewStore(kvStore kv.Store) (*Store, error) {
+	st := &Store{
+		kvStore: kvStore,
+	}
+	return st, st.setup()
+}
+
+func (s *Store) Begin(txType TxType) (*Tx, error) {
+	tx := Tx{
+		close:    make(chan error, 1),
+		response: make(chan error, 1),
+	}
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+	go func() {
+		txFunc := s.kvStore.View
+		if txType == Write {
+			txFunc = s.kvStore.Update
+		}
+		tx.response <- txFunc(context.Background(), func(kvTx kv.Tx) error {
+			tx.Tx = kvTx
+			wg.Done()
```

Comment:
I would consider defering this at the top. A panic in `View` or `Update` or elsewhere in this function could lead Done() not getting called.

Reply:
I see that you need `Done()` to be called before blocking on `<-tx.close`. It is also unlikely that `View` or `Update` would panic as well. But I think this does lead to a complex concurrency dance here.

Any reason we can't stick to the shape of the `kv.Tx` (i.e. `View(fn)` or `Update(fn)`) for the KV backed tenant implementation and then have a separate SQL backed version?
I worry we over complicate things by trying to create some in between storage abstraction which we try to shoe-horn both KV and SQL shaped systems behind.

Reply:
I wanted to use the sql style so when we make the switch the Service layer (the part that uses this) would have very minimal changes. It might have been a bit preemptive.